src/System.Text.Json/src/System/Text/Json/JsonDocument.CustomDb.cs

```diff
@@ -0,0 +1,238 @@
+        //   * 4 bits JsonTokenType
+        //   * 28 bits for the number of rows until the previous value (never 0)
+        //
+        // StartObject
+        // * First int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for token offset
+        // * Second int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for the token length (always 1, effectively unassigned)
+        // * Third int
+        //   * 4 bits JsonTokenType
+        //   * 28 bits for the number of rows until the next value (never 0)
+        //
+        // StartArray
+        // * First int
+        //   * Top bit is unassigned / always clear
+        //   * 31 bits for token offset
+        // * Second int
+        //   * Top bit is set if the array contains other arrays or objects ("complex" types)
+        //   * 31 bits for the number of elements in this array
+        // * Third int
+        //   * 4 bits JsonTokenType
+        //   * 28 bits for the number of rows until the next value (never 0)
+        private struct CustomDb : IDisposable
+        {
+            private const int SizeOrLengthOffset = 4;
+            private const int NumberOfRowsOffset = 8;
+
+            internal int Length;
+            private byte[] _rentedBuffer;
+
+            internal CustomDb(int payloadLength)
+            {
+                // Assume that a token happens approximately every 12 bytes.
+                // int estimatedTokens = payloadLength / 12
+                // now acknowledge that the number of bytes we need per token is 12.
+                // So that's just the payload length.
+                //
+                // Add one token's worth of data just because.
+                int initialSize = DbRow.Size + payloadLength;
+
+                // Stick with ArrayPool's rent/return range if it looks feasible.
+                // If it's wrong, we'll just grow and copy as we would if the tokens
+                // were more frequent anyways.
+                const int OneMegabyte = 1024 * 1024;
+                
+                if (initialSize > OneMegabyte && initialSize <= 4 * OneMegabyte)
+                {
+                    initialSize = OneMegabyte;
+                }
+
+                _rentedBuffer = ArrayPool<byte>.Shared.Rent(initialSize);
+                Length = 0;
+            }
+
+            public void Dispose()
+            {
+                if (_rentedBuffer == null)
+                {
+                    return;
+                }
+
+                // The data in this rented buffer only conveys the positions and
+                // lengths of tokens in a document, but no content; so it does not
+                // need to be cleared.
+                ArrayPool<byte>.Shared.Return(_rentedBuffer);
+                _rentedBuffer = null;
+                Length = 0;
+            }
+
+            internal void TrimExcess()
+            {
+                if (Length <= _rentedBuffer.Length / 2)
```

Comment:
What's the rationale behind this check?

Reply:
The somewhat arbitrary decision of "Is trying to trim worth it?".  The ideal would be ArrayPool.HowBigWouldYouActuallyGiveMeFor(Length) < _rentedBuffer.Length; but we don't have that API.

Given that there's a tradeoff between system memory pressure for effectively dead memory and the time of copying, this seemed like a reasonable starting place.